<launch>
  <group ns="a">
    <include  file="$(find flight_testing)/launch/communication.launch" />

    <!--=============================== Set Flight Origin ================================-->
    <node name="gps_ref" type="gps_ref" pkg="gps_ref">
      <param name="ctrl_mode" value="motor"/> <!--value="first_fix"-->
    </node>

    <!--=========== Convert AscTec GPS data to our coordinate system for tasks ===========-->
    <node name="conv_gps_waypose" type="conversion" pkg="conversion">
      <rosparam param="gps_to_pose">
        [
        [ "task_gps", [ "task_waypose" ] ],
        ]
      </rosparam>
    </node>

    <node name="conv_gps_pose" type="conversion" pkg="conversion">
      <rosparam param="gps_to_pose">
        [
        [ "subject_gps", [ "subject_pose" ] ],
        ]
      </rosparam>
      <param name="latitude_to_meters" value="111050.0"/>
      <param name="longitude_to_meters" value="84389.0"/>
    </node>

    <!--==================================PID Controller==================================--> 
    <node name="pid_ctrl" type="quad_pid" pkg="pose_control">
      <param name="ctrl_rate" value="30.0"/>
    
          <!-- Filtering and control output options -->
          <!-- Fly under GPS control, so the thrust should 
               emulate the barometric pressure control. -->
      <param name="enable_baro_ctrl_mode" value="true" />
          <!-- Flying in barometric pressure control, so collapse the deadband. -->
      <param name="collapse_deadband" value="true" />
          <!-- GPS control requires us to smooth derivative input,
               so enable filtering on this term -->
      <param name="enable_derv_fir_filter" value="true" />

          <!-- PITCH -->
      <param name="k_prop_pitch" value="-0.27"/>
      <param name="k_intg_pitch" value="-0.0"/>
      <param name="k_derv_pitch" value="-0.00"/>
      <param name="pitch_offset" value="0.0"/>
      <param name="pitch_intg_min_sat" value="-1.0"/>
      <param name="pitch_intg_max_sat" value="1.0"/>
     
          <!-- ROLL  -->
      <param name="k_prop_roll" value="-0.27"/>
      <param name="k_intg_roll" value="-0.0"/>
      <param name="k_derv_roll" value="-0.00"/>
      <param name="roll_offset" value="0.0"/>
      <param name="roll_intg_min_sat" value="-1.0"/>
      <param name="roll_intg_max_sat" value="1.0"/>
     
          <!-- YAW  -->
      <param name="k_prop_yaw" value="-0.4"/>
      <param name="k_intg_yaw" value="0.0"/>
      <param name="k_derv_yaw" value="0.0"/>
      <param name="yaw_offset" value="0.0"/>
      <param name="yaw_intg_min_sat" value="0.0"/>
      <param name="yaw_intg_max_sat" value="0.0"/>
     
           <!-- THRUST   -->
      <param name="k_prop_thrust" value="0.15"/>
      <param name="k_intg_thrust" value="0.0"/>
      <param name="k_derv_thrust" value="0.0"/>
      <param name="thrust_offset" value="0.0"/>
      <param name="thrust_intg_min_sat" value="-0.0"/>
      <param name="thrust_intg_max_sat" value="0.0"/>
    </node>

    <!--========================State Specific Position Controllers=====================-->

    <node name="pid_waypose_arb" type="arb_subject_ctrl" pkg="arbitration" output="screen">
      <remap from="shape_shifter" to="pid_waypose"/>
      <rosparam param="state_mappings">
        [
        [ "subject_pose", [ 4 ] ],
        [ "launch_waypose", [ 5 ] ],
        [ "land_waypose", [ 6 ] ],
        [ "task_waypose", [ 8 ] ],
        ]
      </rosparam>
    </node>

    <node name="launch_ctrl" type="delta_ctrl" pkg="pose_control" output="screen">
      <remap from="delta_waypose" to="launch_waypose"/>
      <param name="ctrl_rate" value="30.0"/>
      <param name="delta_rate" value="0.01"/>
      <param name="delta_term" value="translation_z"/>
      <param name="delta_state" value="5"/>
    </node>

    <node name="land_ctrl" type="delta_ctrl" pkg="pose_control" output="screen">
      <remap from="delta_waypose" to="land_waypose"/>
      <param name="ctrl_rate" value="30.0"/>
      <param name="delta_rate" value="-0.2"/>
      <param name="delta_term" value="translation_z"/>
      <param name="delta_state" value="6"/>
    </node>

    <!--======================State Specific PID Controller Input======================-->

    <node name="ctrl_input_arb" type="arb_subject_ctrl" pkg="arbitration" output="screen">
      <remap from="shape_shifter" to="quad_ctrl_input"/>
      <rosparam param="state_mappings">
        [
        [ "null_input", [ 1, 4, 5 ] ],
        [ "ignition_input", [ 2, 3 ] ],
        [ "pid_input", [ 6, 7, 8 ] ],
        ]
      </rosparam>
    </node>

    <node name="null" type="rostopic" pkg="rostopic"
      args="pub -r 30.0 'null_input' collab_msgs/QuadCtrlInput
      '{header: auto, pitch: 0.0, roll: 0.0, yaw: 0.0, thrust: -1.0,
      pitch_ctrl: 1, roll_ctrl: 1, yaw_ctrl: 1, thrust_ctrl: 1,
      altitude_ctrl: 0, gps_ctrl: 0}'"/>

    <node name="ignition" type="rostopic" pkg="rostopic"
      args="pub -r 30.0 'ignition_input' collab_msgs/QuadCtrlInput
      '{header: auto, pitch: 0.0, roll: 0.0, yaw: 1.0, thrust: -1.0,
      pitch_ctrl: 1, roll_ctrl: 1, yaw_ctrl: 1, thrust_ctrl: 1,
      altitude_ctrl: 0, gps_ctrl: 0}'"/>

  </group>
</launch>
